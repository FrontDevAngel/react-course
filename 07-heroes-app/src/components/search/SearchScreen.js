import React, { useMemo } from 'react';
import { useLocation } from 'react-router-dom/cjs/react-router-dom.min';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';

import queryString from 'query-string';
import { getHeroesByName } from '../../selectors/getHeroeByName';

export const SearchScreen = ({ history }) => {
    const location = useLocation();
    // Uso de librerÃ­a de Query String para obtener el parseo del query sin tantas complicaciones ?q=batman
    const { q: query = '' } = queryString.parse(location.search);

    const [values, handleInputsChange] = useForm({ search: query });
    const { search } = values;

    const heroesFiltered = useMemo(() => getHeroesByName(query), [query]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`);
    };

    return (
        <div>
            <h1>Search Screen</h1>

            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Find your hero'
                            className='form-control'
                            name='search'
                            autoComplete='off'
                            onChange={handleInputsChange}
                            value={search}
                        />
                        <button
                            type='submit'
                            className='btn m-1 btn-block btn-outline-primary'>
                            Search...
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />
                    {query === '' && (
                        <div className='alert alert-info'>Search a hero</div>
                    )}

                    {query !== '' && heroesFiltered.length === 0 && (
                        <div className='alert alert-danger'>
                            Couldn't find any {query} hero!
                        </div>
                    )}
                    {heroesFiltered.map((hero) => (
                        <HeroCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </div>
    );
};
