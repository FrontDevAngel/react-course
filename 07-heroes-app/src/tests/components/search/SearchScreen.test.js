import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Test to <SearchScreen />', () => {
    it('should show correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>,
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    it('should show batman and the input with the value of the queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>,
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    it(`should show error if super hero doesnt exist`, () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=raulo']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>,
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe(
            `Couldn't find any raulo hero!`,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('Should call the PUSH to history', () => {
        const history = {
            push: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?']}>
                <Route
                    path='/search'
                    component={() => <SearchScreen history={history} />}
                />
            </MemoryRouter>,
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'batman',
            },
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {},
        });

        expect(history.push).toHaveBeenCalledWith('?q=batman');
    });
});
