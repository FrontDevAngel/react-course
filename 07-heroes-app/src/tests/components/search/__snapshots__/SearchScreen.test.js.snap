// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to <SearchScreen /> should show batman and the input with the value of the queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card"
      >
        <img
          alt="Batman"
          className="card-img-top"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Batman
          </h5>
          <h6
            className="card-subtitle mb-2 text-muted"
          >
            Bruce Wayne
          </h6>
          <p
            className="card-text"
          >
            Detective Comics #27
          </p>
          <p
            className="card-text"
          >
            <small
              className="text-muted"
            >
              DC Comics
            </small>
          </p>
          <a
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            Ver m√°s...
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test to <SearchScreen /> should show correctly with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Test to <SearchScreen /> should show error if super hero doesnt exist 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="raulo"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        Couldn't find any 
        raulo
         hero!
      </div>
    </div>
  </div>
</div>
`;
